blog: TPI - Release post (#3399)
iterative/dvc.org: content/blog/2022-04-27-terraform-provider.md

only the need for specific hardware, but also individual responsibility to shut
down instances when training is complete.

To address the specific needs of machine learning teams, we have built the
Iterative Provider (TPI). TPI is an open-source tool extending the functionality
of Terraform, the world's most widely used multi-cloud provisioning product. The
Iterative Provider enables full lifecycle management of computing resources and
is designed specifically for machine learning pipelines.

## Tailored to Machine Learning Workflows

The Iterative Provider offers a single resource called `iterative_task` which
you can use to configure:

- Your cloud infrastructure
- The steps to perform on the cloud resource, i.e. setting up the environment,
  running the training pipeline, logging metrics, etc.
- The data to be synced back once the training is complete (e.g. a file with
  metrics, a model, plots)

Here’s a “hello world” example of a `main.tf` Terraform configuration file using
the `iterative_task` resource:

```hcl
terraform {
  required_providers { iterative = { source = "iterative/iterative" } }
}
provider "iterative" {}

resource "iterative_task" "example" {

>   cloud   = "aws"
>   region  = "us-east-2"
>   machine = "l+k80"

Reviewer: ```suggestion
  cloud       = "aws"     # or any of: gcp, az, k8s
  machine     = "m"       # medium. Or any of: l, xl, m+k80, xl+v100, ...
  image       = "ubuntu"  # or "nvidia", ...
  region      = "us-east"
  disk_size   = 30        # GB
  spot        = 0         # auto-price. Or -1 to disable, or >0 for hourly USD limit
  timeout     = 60*60     # max 1h before forced termination
```

  storage {
    workdir = "."
    output  = "results"
  }
  script = <<-END
    #!/bin/bash
    sudo apt update
    sudo apt install -y python3-pip
    pip3 install --user -r requirements.txt
    python3 train.py
  END
}
```

Once the training is complete, the Iterative Provider terminates the resource,
so users don't have to worry about spiraling costs from unused machines.

## Configure Once, Bring Everywhere

Once you configure infrastructure and a script that executes your training
pipeline in a Terraform configuration file, you can bring that pipeline anywhere
you want. You can use such a config for ad-hoc training at any stage of your
prototyping process or use it as a job in your preferred CI/CD tool. It will
also store your infrastructure configuration files in a version control system
together with the rest of your project for easier control.

## One Provider to Rule Them All

Whether you prefer Amazon Web Services (AWS), Microsoft Azure, Google Cloud

> Platform (GCP), or Kubernetes, the Iterative Provider has you covered. You can

Reviewer: ```suggestion
Platform (GCP), or Kubernetes (K8s), the Iterative Provider has you covered. You can
```

Reviewer 2: > _One Provider to Rule Them All_

Clever for #devrel, “planeworthy” by and for #engineering. 💍

Reviewer: One Provider to Rule Them All,
One Provider to `spot` them,
One ring to `apply` them all,
and in the darkness `destroy` them.

---

> title: Machine Learning Workloads with Iterative Provider for Terraform

Reviewer: ```suggestion
title: Machine Learning Workloads with Terraform Provider Iterative
```

date: 2022-04-27
description: >
  Today we introduce painless resource orchestration for your machine learning
  projects in conjunction with HashiCorp Terraform.
descriptionLong: >
  Catapult your machine learning projects into any coud and reduce cost with

>   spot instance auto-recovery, while aligning your Data Science  and DevOps

Reviewer: ```suggestion
  spot instance auto-recovery, while aligning your Data Science and DevOps
```

  teams with a simple config file.
picture: 2022-04-27/terraform-provider-iterative.png

> pictureComment: Machine Learning Workloads with Iterative Provider for Terraform

Reviewer: ```suggestion
pictureComment: Machine Learning Workloads with Terraform Provider Iterative
```

author: maria_khalusova
commentsUrl: https://discuss.dvc.org/t/terraform-iterative-provider-release-blog-post/1171
tags:
  - MLOps
  - Terraform
  - Git
  - Resource orchestation
  - Spot Instance management
---

The requirements for Machine Learning (ML) infrastructure are becoming
increasingly complex. Training large models often requires specialized hardware
(GPUs, TPUs) which involves moving the whole training process onto cloud
machines, setting up environments and synchronizing data. For teams that want to

> leverage spot instances, the setup becomes even more complex - they need to make

Reviewer: ```suggestion
leverage spot instances, the setup becomes even more complex -- they need to make
```

sure the training progress is not lost during spot instance recovery. This is

> time-consuming, and requires expertise in both DevOps and Machine Learning. On
> top of that, the costs of training in a cloud environment can be high due to not
> only the need for specific hardware, but also individual responsibility to shut
> down instances when training is complete.

Reviewer: ```suggestion
time-consuming, and requires expertise in both DevOps and Machine Learning.
Additionally, training in a cloud environment can incur high costs due to the
need for expensive hardware, as well as users forgetting to shutdown instances
when training is complete.
```

> To address the specific needs of machine learning teams, we have built Terraform
> Provider Iterative (TPI). TPI is an open-source tool extending the functionality

Reviewer: ```suggestion
To address the specific needs of machine learning teams, we have built
[Terraform Provider Iterative (TPI)](https://github.com/iterative/terraform-provider-iterative).
TPI is an open-source tool extending the functionality
```

> To address the specific needs of machine learning teams, we have built the
> Iterative Provider (TPI). TPI is an open-source tool extending the functionality

Reviewer: ```suggestion
To address the specific needs of machine learning teams, we have built Terraform
Provider Iterative (TPI). TPI is an open-source tool extending the functionality
```

of Terraform, the world's most widely used multi-cloud provisioning product. The
Iterative Provider enables full lifecycle management of computing resources and
is designed specifically for machine learning pipelines.

## Tailored to Machine Learning Workflows

The Iterative Provider offers a single resource called `iterative_task` which
you can use to configure:

- Your cloud infrastructure
- The steps to perform on the cloud resource, i.e. setting up the environment,
  running the training pipeline, logging metrics, etc.
- The data to be synced back once the training is complete (e.g. a file with
  metrics, a model, plots)

Here’s a “hello world” example of a `main.tf` Terraform configuration file using
the `iterative_task` resource:

```hcl
terraform {
  required_providers { iterative = { source = "iterative/iterative" } }
}
provider "iterative" {}

resource "iterative_task" "example" {
  cloud       = "aws"     # or any of: gcp, az, k8s
  machine     = "m"       # medium. Or any of: l, xl, m+k80, xl+v100, ...
  image       = "ubuntu"  # or "nvidia", ...
  region      = "us-west" # or us-west, eu-east, ...
  disk_size   = 30        # GB

>   spot        = 0         # auto-price. Default -1 to disable, or >0 for hourly USD limit

Reviewer: ```suggestion
  spot        = 0         # auto-price. Default -1 to disable or >0 for hourly USD limit
```

>   region      = "us-east"
>   disk_size   = 30        # GB
>   spot        = 0         # auto-price. Or -1 to disable, or >0 for hourly USD limit
>   timeout     = 60*60     # max 1h before forced termination

Reviewer: ```suggestion
  region      = "us-west" # or us-west, eu-east, ...
  disk_size   = 30        # GB
  spot        = 0         # auto-price. Default -1 to disable, or >0 for hourly USD limit
  timeout     = 24*60*60  # max 24h before forced termination
```

  storage {
    workdir = "."
    output  = "results"
  }
  script = <<-END
    #!/bin/bash
    sudo apt update
    sudo apt install -y python3-pip
    pip3 install --user -r requirements.txt
    python3 train.py
  END
}
```

Once the training is complete, the Iterative Provider terminates the resource,
so users don't have to worry about spiraling costs from unused machines.

## Configure Once, Bring Everywhere

Once you configure infrastructure and a script that executes your training
pipeline in a Terraform configuration file, you can bring that pipeline anywhere
you want. You can use such a config for ad-hoc training at any stage of your

> prototyping process or use it as a job in your preferred CI/CD tool. It will

Reviewer: ```suggestion
prototyping process or use it as a job in your preferred CI/CD tool. You can
```

also store your infrastructure configuration files in a version control system
together with the rest of your project for easier control.

## One Provider to Rule Them All

Whether you prefer Amazon Web Services (AWS), Microsoft Azure, Google Cloud
Platform (GCP), or Kubernetes (K8s), the Iterative Provider has you covered. You

> can configure compute resources from these with a single provider, using
> [common machine types](https://registry.terraform.io/providers/iterative/iterative/latest/docs/resources/task#machine-type).
> These are going to be roughly similar between different cloud vendors. Not only
> does this significantly simplify infrastructure configuration, it also makes it
> easy to migrate from one cloud to another with only minor adjustments.

Reviewer: ```suggestion
can configure compute resources from these with a unified API, using
[common machine types](https://registry.terraform.io/providers/iterative/iterative/latest/docs/resources/task#machine-type)
that are the same across all cloud vendors. This significantly simplifies
infrastructure configuration and makes it easy to migrate from one cloud to another
by changing just one line of code.
```

## Costs Optimization

The Iterative Provider helps with cloud compute cost optimization in two major
ways. First, upon completion of your script, the instance is automatically
terminated. This helps to avoid accumulating costs due to abandoned resources.
Second, you can leverage the cost-saving power of spot instances to train your

> models without losing any progress!

Reviewer: ```suggestion
models without losing any progress! TPI recovers the working directory and
respawns interrupted/preempted instances for you.
```

---
title: Machine Learning Workloads with Terraform Provider Iterative
date: 2022-04-27
description: >
  Today we introduce painless resource orchestration for your machine learning
  projects in conjunction with HashiCorp Terraform.
descriptionLong: >
  Catapult your machine learning projects into any coud and reduce cost with
  spot instance auto-recovery, while aligning your Data Science and DevOps teams
  with a simple config file.
picture: 2022-04-27/terraform-provider-iterative.png
pictureComment: Machine Learning Workloads with Terraform Provider Iterative
author: maria_khalusova
commentsUrl: https://discuss.dvc.org/t/terraform-iterative-provider-release-blog-post/1171
tags:

>   - MLOps
>   - Terraform
>   - Git
>   - Resource orchestation
>   - Spot Instance management

Reviewer: ```suggestion
  - MLOps
  - Terraform
  - Git
  - Cloud orchestration
  - Spot instance management
```

top of that, the costs of training in a cloud environment can be high due to not
only the need for specific hardware, but also individual responsibility to shut
down instances when training is complete.

To address the specific needs of machine learning teams, we have built the
Iterative Provider (TPI). TPI is an open-source tool extending the functionality
of Terraform, the world's most widely used multi-cloud provisioning product. The
Iterative Provider enables full lifecycle management of computing resources and
is designed specifically for machine learning pipelines.

## Tailored to Machine Learning Workflows

The Iterative Provider offers a single resource called `iterative_task` which
you can use to configure:

- Your cloud infrastructure
- The steps to perform on the cloud resource, i.e. setting up the environment,
  running the training pipeline, logging metrics, etc.
- The data to be synced back once the training is complete (e.g. a file with
  metrics, a model, plots)

Here’s a “hello world” example of a `main.tf` Terraform configuration file using
the `iterative_task` resource:

```hcl
terraform {

>   required_providers {
>     iterative = {
>       source = "iterative/iterative"
>     }
>   }
> }
> 
> provider "iterative" {}

Reviewer: ```suggestion
  required_providers { iterative = { source = "iterative/iterative" } }
}
provider "iterative" {}
```

resource "iterative_task" "example" {
  cloud   = "aws"
  region  = "us-east-2"
  machine = "l+k80"

>   script = <<-END
>     #!/bin/bash
>     sudo apt update
>     sudo apt install -y python3-pip
>     pip3 install -r requirements.txt
>     python3 src/train.py
>   END
> 
>   storage {
>     workdir = "."
>     output  = "results"
>   }

Reviewer: ```suggestion
  storage {
    workdir = "."
    output  = "results"
  }
  script = <<-END
    #!/bin/bash
    sudo apt update
    sudo apt install -y python3-pip
    pip3 install --user -r requirements.txt
    python3 train.py
  END
```
